cocos/core/geometry/spline.ts

```diff
@@ -0,0 +1,292 @@
+            this._knots[i] = new Vec3(knots[i]);
+        }
+    }
+
+    public static create (mode: SplineMode, knots: Vec3[] = []) {
+        return new Spline(mode, knots);
+    }
+
+    public static clone (s: Spline) {
+        return new Spline(s.getMode(), s.getKnots());
+    }
+
+    public static copy (out: Spline, s: Spline) {
+        out._mode = s.getMode();
+        out._knots = [];
+
+        for (let i = 0; i < s.getKnots().length; i++) {
+            out._knots[i] = new Vec3(s.getKnots()[i]);
+        }
+
+        return out;
+    }
+
+    get type () {
+        return this._type;
+    }
+
+    public getMode () {
+        return this._mode;
+    }
+
+    public setMode (value) {
+        this._mode = value;
+    }
+
+    public getKnots () {
+        return this._knots;
+    }
+
+    public setKnots (value) {
+        this._knots = [];
+
+        for (let i = 0; i < value.length; i++) {
+            this._knots[i] = new Vec3(value[i]);
+        }
+    }
+
+    public clearKnots() { 
+        this._knots = [];
+    }
+
+    public getKnotCount() { 
+        return this._knots.length; 
+    }
+
+    public addKnot(knot: Vec3) { 
+        this._knots.push(new Vec3(knot)); 
+    }
+
+    public insertKnot(index: number, knot: Vec3) {
+        const item = new Vec3(knot);
+        if (index >= this._knots.length) {
+            this._knots.push(item);
+            return;
+        }
+    
+        this._knots.splice(index, 0, item);
+    }
+
+    public removeKnot(index: number) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        this._knots.splice(index, 1);
+    }
+
+    public setKnot(index: number, knot: Vec3) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        this._knots[index].set(knot);
+    }
+
+    public getKnot(index: number) {
+        assertIsTrue(index < this._knots.length, "Spline: invalid index");
+
+        return this._knots[index];
+    }
+
+    // get a point at t with repect to the `index` segment of curve or the whole curve.
+    public getPoint(t: number, index: number = SPLINE_WHOLE_INDEX) {
+        t = clamp(t, 0.0, 1.0);
+
+        const segments = this.getSegments();
+        if (segments == 0) {
+            return new Vec3(0.0, 0.0, 0.0);
+        }
+
+        if (index == SPLINE_WHOLE_INDEX) {
+            const deltaT = 1.0 / segments;
+
+            index = Math.floor(t / deltaT);
+            t     = (t % deltaT) / deltaT;
+        }
+
+        if (index >= segments) {
+            return this._knots[this._knots.length - 1];
+        }
+
+        switch (this._mode) {
+            case SplineMode.LINEAR:
+                return Spline.calcLinear(this._knots[index], this._knots[index + 1], t);
+            case SplineMode.BEZIER:
+                return Spline.calcBezier(this._knots[index * 4], this._knots[index * 4 + 1], this._knots[index * 4 + 2], this._knots[index * 4 + 3], t);
```

Comment:
How to make sure `index * 4 + 3` will not exceed the size?

Reply:
Bezier mode requires to add multiple of 4 knots to spline.